### PR TITLE
Adding cube structure init events

### PR DESCRIPTION
**Pull request by Foghrye4**

### CODE REVIEW

File: src/main/java/io/github/opencubicchunks/cubicchunks/cubicgen/customcubic/structure/CubicCaveGenerator.java [hunk 1]
```diff
@@ -32,6 +32,7 @@
 import static net.minecraft.util.math.MathHelper.sin;
 
 import io.github.opencubicchunks.cubicchunks.api.worldgen.CubePrimer;
+import io.github.opencubicchunks.cubicchunks.api.worldgen.CubicStructureGenerator;
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/cubicgen/customcubic/structure/CubicCaveGenerator.java:35]: why?

File: src/main/java/io/github/opencubicchunks/cubicchunks/cubicgen/customcubic/structure/CubicRavineGenerator.java [hunk 2]
```diff
@@ -31,6 +31,7 @@
 
 import io.github.opencubicchunks.cubicchunks.api.util.Coords;
 import io.github.opencubicchunks.cubicchunks.api.worldgen.CubePrimer;
+import io.github.opencubicchunks.cubicchunks.api.worldgen.CubicStructureGenerator;
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/cubicgen/customcubic/structure/CubicRavineGenerator.java:34]: same here

File: src/main/java/io/github/opencubicchunks/cubicchunks/cubicgen/customcubic/CustomTerrainGenerator.java [hunk 3]
```diff
@@ -111,8 +113,17 @@ private CustomTerrainGenerator(World world, BiomeProvider biomeProvider, CustomG
...
+        InitCubicStructuteGeneratorEvent caveEvent = new InitCubicStructuteGeneratorEvent(EventType.CAVE, new CubicCaveGenerator());
+        InitCubicStructuteGeneratorEvent strongholdsEvent = new InitCubicStructuteGeneratorEvent(EventType.STRONGHOLD, new CubicStrongholdGenerator(conf));
+        InitCubicStructuteGeneratorEvent ravineEvent = new InitCubicStructuteGeneratorEvent(EventType.RAVINE, new CubicRavineGenerator(conf));
+
+        MinecraftForge.TERRAIN_GEN_BUS.post(caveEvent);
+        MinecraftForge.TERRAIN_GEN_BUS.post(strongholdsEvent);
+        MinecraftForge.TERRAIN_GEN_BUS.post(ravineEvent);
+        
+        this.caveGenerator = (CubicCaveGenerator) caveEvent.getNewGen();
+        this.strongholds = (CubicFeatureGenerator) strongholdsEvent.getNewGen();
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/cubicgen/customcubic/CustomTerrainGenerator.java:125]: Does it need to be this specific type? Those classes aren't in API so there is no way to actually replace them.

**Review by Niko-sk2x — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/opencubicchunks/cubicchunks/cubicgen/customcubic/CustomTerrainGenerator.java (hunk 3 above)
Comment (Foghrye4) [reply src/main/java/io/github/opencubicchunks/cubicchunks/cubicgen/customcubic/CustomTerrainGenerator.java:125]: Strongholds generator call `generateStructure(..)` and `getClosestStrongholdPos(...)`, which is not presented in `CubicStructureGenerator.class`. I can move those functions to `CubicStructureGenerator.class` which is moved to CC API.

Comment (Niko-sk2x) [reply src/main/java/io/github/opencubicchunks/cubicchunks/cubicgen/customcubic/CustomTerrainGenerator.java:125]: This doesn't appear to atch method names in 1.12.x vanilla. Can you see how it's done in vanilla?

Comment (Niko-sk2x) [reply src/main/java/io/github/opencubicchunks/cubicchunks/cubicgen/customcubic/CustomTerrainGenerator.java:125]: Also looks like CubicFeatureGenerator may need to be in API. (after renames to match vanilla)

Comment (Foghrye4) [reply src/main/java/io/github/opencubicchunks/cubicchunks/cubicgen/customcubic/CustomTerrainGenerator.java:125]: @Barteks2x 
In Vanilla it looks like this:
```            caveGenerator = net.minecraftforge.event.terraingen.TerrainGen.getModdedMapGen(caveGenerator, net.minecraftforge.event.terraingen.InitMapGenEvent.EventType.CAVE);```
https://github.com/MinecraftForge/MinecraftForge/blob/1.12.x/patches/minecraft/net/minecraft/world/gen/ChunkGeneratorOverworld.java.patch#L8

Comment (Niko-sk2x) [reply src/main/java/io/github/opencubicchunks/cubicchunks/cubicgen/customcubic/CustomTerrainGenerator.java:125]: I wasn't asking what the event looks like, but the code for `getClosestStrongholdPos`. That method got renamed in some version.

Comment (Foghrye4) [reply src/main/java/io/github/opencubicchunks/cubicchunks/cubicgen/customcubic/CustomTerrainGenerator.java:125]: ```            return this.strongholdGenerator.getNearestStructurePos(worldIn, position, findUnexplored);```
While `getNearestStructurePos(...)` is a member of `MapGenStructure.class`

Comment (Niko-sk2x) [reply src/main/java/io/github/opencubicchunks/cubicchunks/cubicgen/customcubic/CustomTerrainGenerator.java:125]: So rename it in `CubicFeatureGenerator` to match vanilla and move this class to API.

**Comment by Niko-sk2x** [comment]
Merged manually (+ my own changes) in https://github.com/OpenCubicChunks/CubicWorldGen/commit/847c08c39049318afa9fa03ecb7e0e0b3f2875a9

**Pull request closed without merging**